### ISSUE TITLE
Image Integrity Scan 

### ISSUE THREAD

**Issue by moshloop** [goharbor/harbor#10846]
It seems that images can sometimes be partially created (missing links or 0B files) - it would be great if a scheduled job can be run to remove and/or fix these images as it can cause replication to fail.

**Comment by ninjadq** [comment]
could please provide more details and log?

**Comment by steven-zou** [comment]
As it mostly likes a requirement, and then let @xaleeks decide if it is reasonable. Thanks.

**Comment by moshloop** [comment]
I think this might be related to https://github.com/goharbor/harbor/issues/9295 - I don't see any errors in the logs for the replication job that fails completely (There are other errors for replication jobs that fail partially) 

e.g. 

```
harbor-harbor-jobservice-7565fc7f6c-2t9f6 jobservice 2020-02-27T08:58:19Z [ERROR] [/jobservice/runner/redis.go:64]: |@_@| Job 'REPLICATION:ac41b87bb56e3aa25eaab313' exit with error: http error: code 404, message {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":{"Name":"dn","Revision":"sha256:6eb51ffc4856500c52070eb7721b65048711feadc5f065e91aaa16f04f72109e"}}]}
```

**Comment by xaleeks** [comment]
so you are connecting to gitlab? @moshloop

**Comment by moshloop** [comment]
No - this is a 1.10.0 to 1.10.0 habor to harbor replication

**Comment by xaleeks** [comment]
So I'm a bit confused, why is this called image integrity scan?  Or is it a replication that is producing these 0kb files.  That' something we need to fix. @moshloop

**Comment by moshloop** [comment]
The replication fails to even start because there are images / tags that are corrupted - The purpose of this scan / job is to cleanup these images and tags so that the replication job can a) start and b) complete with 100% success rate consistently

**Comment by michmike** [comment]
@moshloop thanks for raising this issue.
Is my understanding accurate here
* In some situations, pushing an image into harbor fails for whatever reason and you end up with a 0kb corrupt image in Harbor
* replication then fails because of this 0kb artifact

I agree that we should make replication more resilient where it should log errors/alerts when it can't replicate an artifact, but it should not be stuck on it

How often do you get these corrupted images? Can your CI/CD process, or whatever is pushing the images, try again and reupload the same image+tag so that it replaces the corrupt one?

**Comment by moshloop** [comment]
> In some situations, pushing an image into harbor fails for whatever reason and you end up with a 0kb corrupt image in Harbor

This is not necessarily a 0kb image, it could be missing links, digests etc..

> How often do you get these corrupted images?

This is not often, less than < 1% 

> Can your CI/CD process, or whatever is pushing the images, try again and re-upload the same image+tag so that it replaces the corrupt one?

If we could supply teams with a list of corrupted image+tags that an image integrity scan could produce, then potentially yes, or we could delete them

**Comment by xaleeks** [comment]
We believe this is upstream distribution issue.  We've only come across it a few times and it's hard to reproduce but very similar to the issues raised on distribution repo.  In terms of mitigating its impact on future operations like replications, we are investigate how to clean these off the registry or fail gracefully in our resolver.

**Comment by michmike** [comment]
@xaleeks another thing we can do is when an image is incorrectly pushed into harbor, we can have a webhook option that an operator can subscribe too. can you please file an issue on that?

**Comment by xaleeks** [comment]
@reasonerjt assuming this is the 0kb issue we've come across before, is there a way to detect it as soon as it surfaces on the 2.0 registry.  on 2.0, is an artifact created here or something created in db that we can flag?  If it is, we can clean up automatically bypassing the need for a webhook.

**Comment by xaleeks** [comment]
This should not be an issue on Harbor 2.0 since we will check manifests before insertion into DB so they are not carelessly generated anymore.  Right now there is no quick patch for since this mechanism doesn't exist for Harbor 1.x. I'm keeping this open to track if this comes up again on v2.0

**Comment by xaleeks** [comment]
Let's make sure this is indeed fixed on the 2.0 like we anticipated.  @wy65701436 @heww

**Comment by reasonerjt** [comment]
We can make sure there won't be 0KB issue when pushing image to Harbor after 2.0.0, but we still need the workaround for upgraded instance where there are 0KB images.

**Comment by michmike** [comment]
@xaleeks can we confirm we fixed the issue where Harbor 2.0 won't let you push `corrupted` images?

**Comment by xaleeks** [comment]
I spoke to Daniel and this shouldn't happening anymore in 2.0 for reason stated above, provided that this is the same bug detailed here and it looks like it is.
